src/python/pants/backend/python/rules/python_create_binary.py

```diff
@@ -1,38 +1,79 @@
 # Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
 
+from dataclasses import dataclass
+from typing import Optional
+
 from pants.backend.python.rules.pex import Pex
 from pants.backend.python.rules.pex_from_target_closure import CreatePexFromTargetClosure
+from pants.backend.python.rules.targets import EntryPoint, PythonBinarySources
+from pants.backend.python.rules.targets import targets as python_targets
 from pants.backend.python.targets.python_binary import PythonBinary
+from pants.build_graph.address import Address
 from pants.engine.addressable import Addresses
 from pants.engine.legacy.structs import PythonBinaryAdaptor
+from pants.engine.parser import HydratedStruct
 from pants.engine.rules import UnionRule, rule
 from pants.engine.selectors import Get
+from pants.engine.target import SourcesRequest, SourcesResult, Target, hydrated_struct_to_target
 from pants.rules.core.binary import BinaryTarget, CreatedBinary
-from pants.rules.core.determine_source_files import AllSourceFilesRequest, SourceFiles
+from pants.rules.core.strip_source_roots import SourceRootStrippedSources, StripSnapshotRequest
+
+
+@dataclass(frozen=True)
+class PythonBinaryFields:
```

Comment:
We can't use the name `PythonBinary` because that's already used by the target type `PythonBinary`. 

Also, the point of this is that we don't care at all if it's a `python_binary` or not. The sole thing we care about is that the target has the fields `PythonBinarySources` and `EntryPoint` registered.

Reply:
It sounds like rather than something that folks will need to create as boilerplate in the long term, one of the reasons this exists is that it dodges the need for https://github.com/pantsbuild/pants/issues/7490 ? Because otherwise a "caller" could `await Get[CreatedBinary]((Address(..), PythonBinarySources(..), EntryPoint(..)))` ...?

Reply:
Not necessarily, because we need to support optional fields + opt-out fields. I don't think we could use #7490 to ergonomically say `Maybe(ZipSafe)` without new engine abstractions. In contrast, using something like `PythonBinaryFields` allows us to do this:

```python
def is_valid_target(tgt: Target) -> bool:
  return tgt.has_fields([EntryPoint, PythonBinarySources]) and not tgt.has_field(NotPythonBinary)
```